src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsTypedHandler.java

```diff
@@ -0,0 +1,208 @@
+import com.intellij.openapi.editor.CaretModel;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.fileTypes.FileType;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LSPIJEditorUtils;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.lsp4ij.LanguageServiceAccessor;
+import com.redhat.devtools.lsp4ij.client.features.EditorBehaviorFeature;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Improves editor-typed handling of nested quote characters and statement terminators.
+ */
+public class LSPEditorImprovementsTypedHandler extends TypedHandlerDelegate {
+
+    @NotNull
+    @Override
+    public Result beforeCharTyped(char charTyped,
+                                  @NotNull Project project,
+                                  @NotNull Editor editor,
+                                  @NotNull PsiFile file,
+                                  @NotNull FileType fileType) {
+        if (LSPIJEditorUtils.isSupportedAbstractFileTypeOrTextMateFile(file)) {
+            // String literal improvements
+            if (LSPIJEditorUtils.isQuoteCharacter(file, charTyped) &&
+                EditorBehaviorFeature.enableStringLiteralImprovements(file) &&
+                handleNestedQuote(file, editor, charTyped)) {
+                return Result.STOP;
+            }
+            // Statement terminator improvements
+            else if (EditorBehaviorFeature.enableStatementTerminatorImprovements(file) &&
+                     isStatementTerminatorCharacter(file, charTyped) &&
+                     handleStatementTerminator(file, editor)) {
+                return Result.STOP;
+            }
+        }
+
+        return Result.CONTINUE;
+    }
+
+    private boolean handleNestedQuote(@NotNull PsiFile file,
+                                      @NotNull Editor editor,
+                                      char quote) {
+        Document document = editor.getDocument();
+        CharSequence documentChars = document.getCharsSequence();
+        int offset = editor.getCaretModel().getOffset();
+
+        // Try to find a string literal that contains the current offset
+        String stringLiteral = getStringLiteral(file, documentChars, offset);
+        if (StringUtil.isNotEmpty(stringLiteral)) {
+            char stringLiteralQuote = stringLiteral.charAt(0);
+            // Single quote within double-quoted string or vice-versa
+            // Escaped single quote within single-quoted string or vice-versa
+            if ((stringLiteralQuote != quote) || ((offset > 0) && (documentChars.charAt(offset - 1) == '\\'))) {
```

Comment:
If inserting a quote character different from the one used for the containing string literal, or if inserting an escaped quote character, just insert the character and advance the offset so that it's effectively typed, then return `true` and the `typedHandler` will return `STOP` so that no further processing of the typed character occurs. That way there won't be any attempt to insert a paired quote character or advance the caret without inserting if escaping just before the string literal's close quote character.